### ISSUE TITLE
distances() on particular graph causes segfault

### ISSUE THREAD

**Issue by HughParsonage** [igraph/rigraph#467]
Reprex::

```r
library(data.table)
library(igraph)  # 1.2.6

set.seed(1)
TestEdges <- data.table(orig = sample(1:250e3, size = 2e6, replace = TRUE), 
                        dest = sample(1:250, size = 2e6, replace = TRUE), 
                        key = "orig,dest")
GraphZ <- igraph::graph_from_data_frame(TestEdges, directed = FALSE)
Distances <- igraph::distances(GraphZ)
```

**Comment by gaborcsardi** [comment]
Is this supposed to return a 2e6 x 2e6 matrix of doubles? You'll need about 29 TeraBytes of memory to hold that.

**Comment by HughParsonage** [comment]
Not quite! It was just an easy to reproduce reprex. Thought the segfault warranted attention regardless of practicality.

Nonetheless, with `size = 57e3` I still reproduce the segfault, even though I have ample RAM for that. (Noticeably `size = 47e3` does not realiably segfault).

**Comment by HughParsonage** [comment]
Windows. 128GB

**Comment by szhorvat** [comment]
Windows is a bit problematic, because the `long int` type is only 32 bits wide, even on 64-bit Windows systems. igraph uses `long int` extensively for internal indexing.  Your matrix is much too large to be indexable with 32-bit integers, so I am not surprised that things went wrong.

Currently the igraph C library is being refactored to improve on the situation, but it will take some time before the improvements will make it into the R interface.

**Comment by szhorvat** [comment]
I expect the largest feasible _square_ matrix size to be `sqrt(2^31-1)` which is 46341.

**Comment by szhorvat** [comment]
Closing in favour of #405, they will be solved together in the C core for 2.0.
